[SPARK-14500] [ML] Accept Dataset[_] instead of DataFrame in MLlib APIs

mengxr opened the pull request:
## What changes were proposed in this pull request?

This PR updates MLlib APIs to accept `Dataset[_]` as input where `DataFrame` was the input type. This PR doesn't change the output type. In Java, `Dataset[_]` maps to `Dataset<?>`, which includes `Dataset<Row>`. Some implementations were changed in order to return `DataFrame`. Tests and examples were updated. Note that this is a breaking change for subclasses of Transformer/Estimator.

Lol, we don't have to rename the input argument, which has been `dataset` since Spark 1.2.

TODOs:
- [x] update MiMaExcludes (seems all covered by explicit filters from SPARK-13920)
- [x] Python
- [x] add a new test to accept Dataset[LabeledPoint]
- [x] remove unused imports of Dataset
## How was this patch tested?

Exiting unit tests with some modifications.

cc: @rxin @jkbradley

mengxr committed:
accept Dataset[_] instead of DataFrame in MLlib

mengxr committed:
fix compile

mengxr committed:
fix tests

mengxr committed:
Merge remote-tracking branch 'apache/master' into SPARK-14500

mengxr committed:
fix examples

SparkQA commented:
**[Test build #55433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55433/consoleFull)** for PR 12274 at commit [`3f765dd`](https://github.com/apache/spark/commit/3f765dd75df8afb444bb433a209cf4237f584b29).

SparkQA commented:
**[Test build #55433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55433/consoleFull)** for PR 12274 at commit [`3f765dd`](https://github.com/apache/spark/commit/3f765dd75df8afb444bb433a209cf4237f584b29).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55433/
Test FAILed.

mengxr committed:
fix scalastyle

SparkQA commented:
**[Test build #55434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55434/consoleFull)** for PR 12274 at commit [`42bda51`](https://github.com/apache/spark/commit/42bda51968ba6ea0cef6df41451c025f94b63bd3).

mengxr committed:
remove unused imports

SparkQA commented:
**[Test build #55435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55435/consoleFull)** for PR 12274 at commit [`1a62563`](https://github.com/apache/spark/commit/1a62563960d3571379ba2e25f5a98de7f64dc78f).

SparkQA commented:
**[Test build #55434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55434/consoleFull)** for PR 12274 at commit [`42bda51`](https://github.com/apache/spark/commit/42bda51968ba6ea0cef6df41451c025f94b63bd3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55434/
Test PASSed.

SparkQA commented:
**[Test build #55435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55435/consoleFull)** for PR 12274 at commit [`1a62563`](https://github.com/apache/spark/commit/1a62563960d3571379ba2e25f5a98de7f64dc78f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55435/
Test PASSed.

mengxr committed:
add a test in glm for Dataset[LabeledPoint] input

SparkQA commented:
**[Test build #55440 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55440/consoleFull)** for PR 12274 at commit [`757e413`](https://github.com/apache/spark/commit/757e41354043c77f82d29a568500a3f280bc1c15).

SparkQA commented:
**[Test build #55440 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55440/consoleFull)** for PR 12274 at commit [`757e413`](https://github.com/apache/spark/commit/757e41354043c77f82d29a568500a3f280bc1c15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55440/
Test PASSed.

rxin commented:
High level change LGTM!

jaceklaskowski commented on mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala:
```diff
@@ -124,7 +124,7 @@ class Pipeline @Since("1.4.0") (
    * @return fitted pipeline
    */
   @Since("1.2.0")
```
What about `@Since` then? It's a different method, isn't it?

jaceklaskowski commented on mllib/src/main/scala/org/apache/spark/ml/Predictor.scala:
```diff
@@ -171,18 +171,18 @@ abstract class PredictionModel[FeaturesType, M <: PredictionModel[FeaturesType,
    * @return transformed dataset with [[predictionCol]] of type [[Double]]
    */
-  override def transform(dataset: DataFrame): DataFrame = {
+  override def transform(dataset: Dataset[_]): DataFrame = {
```
What about the return type? If I want to chain transformers by `andThen`, i.e. `(tok.transform _).andThen(hashTF.transform)`, won't `DataFrame` be an issue? Why do we keep `DataFrame` as the return type?

jaceklaskowski commented on mllib/src/main/scala/org/apache/spark/mllib/regression/impl/GLMRegressionModel.scala:
```diff
@@ -23,7 +23,7 @@ import org.json4s.jackson.JsonMethods._
 import org.apache.spark.mllib.linalg.Vector
 import org.apache.spark.mllib.util.Loader
-import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+import org.apache.spark.sql.{DataFrame, Dataset, Row, SQLContext}
```
I'm surprised you need this change without any other changes in the file? How could that be?

jaceklaskowski commented on mllib/src/main/scala/org/apache/spark/mllib/tree/model/DecisionTreeModel.scala:
```diff
@@ -32,7 +32,7 @@ import org.apache.spark.mllib.tree.configuration.{Algo, FeatureType}
 import org.apache.spark.mllib.util.{Loader, Saveable}
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+import org.apache.spark.sql.{DataFrame, Dataset, Row, SQLContext}
```
Again, no other change in the file and you need the import?

mengxr committed:
remove other unused imports

mengxr committed:
update since versions

SparkQA commented:
**[Test build #55472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55472/consoleFull)** for PR 12274 at commit [`6409245`](https://github.com/apache/spark/commit/64092459b3ea3ff2f56a9d8e12a11efe7936e6b8).

mengxr commented:
@jaceklaskowski Thanks for reviewing! I updated since versions and removed unused imports. The return type should remain `DataFrame`, i.e., `Dataset[Row]`, which is compatible with `Dataset[_]` when used as input.

SparkQA commented:
**[Test build #55473 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55473/consoleFull)** for PR 12274 at commit [`2c2a4bf`](https://github.com/apache/spark/commit/2c2a4bfbe9955950ea67ec7aa2a252390fb27c54).

SparkQA commented:
**[Test build #55472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55472/consoleFull)** for PR 12274 at commit [`6409245`](https://github.com/apache/spark/commit/64092459b3ea3ff2f56a9d8e12a11efe7936e6b8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55472/
Test PASSed.

SparkQA commented:
**[Test build #55473 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55473/consoleFull)** for PR 12274 at commit [`2c2a4bf`](https://github.com/apache/spark/commit/2c2a4bfbe9955950ea67ec7aa2a252390fb27c54).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55473/
Test PASSed.

mengxr commented:
Merged into master.
